src/libopensc/card-dnie.c

```diff
@@ -729,6 +729,62 @@ int dnie_match_card(struct sc_card *card)
 	LOG_FUNC_RETURN(card->ctx, result);
 }
 
+static int dnie_sm_free_wrapped_apdu(struct sc_card *card,
+		struct sc_apdu *plain, struct sc_apdu **sm_apdu)
+{
+	struct sc_context *ctx = card->ctx;
+	int rv = SC_SUCCESS;
+
+	LOG_FUNC_CALLED(ctx);
+	if (!sm_apdu)
+		LOG_FUNC_RETURN(ctx, SC_ERROR_INVALID_ARGUMENTS);
+	if (!(*sm_apdu))
+		LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+
+	plain->resp = (*sm_apdu)->resp;
+	plain->resplen = (*sm_apdu)->resplen;
+	plain->sw1 = (*sm_apdu)->sw1;
+	plain->sw2 = (*sm_apdu)->sw2;
```

Comment:
That is what I am trying to do.
I haven't been able to do it in one shot yet.
It might look easy to do, maybe it is and I just haven't found the way yet. If I move the encoding and the surrounding APDU modifications to dnie_sm_get_wrapped_apdu I get APDU validation errors that I haven't been able to sort out so far. If I move the decoding to dnie_sm_free_wrapped_apdu, I get decoding errors.
I need to work a bit on how to make the current transformations in a different place so that they have the same effect.
I certainly do want to remove all this historic burden as you call it.